src/embedded_flash/syscall.rs

```diff
@@ -157,83 +192,127 @@ impl Storage for SyscallStorage {
+    /// Provides access to the other upgrade partition and metadata if available.
+    ///
+    /// # Errors
+    ///
+    /// Returns `CustomError` if any of the following conditions do not hold:
+    /// - The page size is a power of two.
+    /// - The storage slices are page-aligned.
+    /// - There are not partition or metadata slices.
+    /// Returns a `NotAligned` error if partitions or metadata ranges are
+    /// - not exclusive or,
+    /// - not consecutive.
+    pub fn new() -> StorageResult<SyscallUpgradeStorage> {
+        let mut locations = SyscallUpgradeStorage {
+            page_size: get_info(command_nr::get_info_nr::PAGE_SIZE, 0)?,
+            partition: ModRange::new_empty(),
+            metadata: ModRange::new_empty(),
+        };
+        if !locations.page_size.is_power_of_two() {
             return Err(StorageError::CustomError);
         }
-        Ok(())
+        for i in 0..memop(memop_nr::STORAGE_CNT, 0)? {
+            let storage_type = memop(memop_nr::STORAGE_TYPE, i)?;
+            match storage_type {
+                storage_type::PARTITION | storage_type::METADATA => (),
+                _ => continue,
+            };
+            let storage_ptr = memop(memop_nr::STORAGE_PTR, i)?;
+            let storage_len = memop(memop_nr::STORAGE_LEN, i)?;
+            if !locations.is_page_aligned(storage_ptr) || !locations.is_page_aligned(storage_len) {
+                return Err(StorageError::CustomError);
+            }
+            let range = ModRange::new(storage_ptr, storage_len);
+            match storage_type {
+                storage_type::PARTITION => {
+                    locations.partition = locations
+                        .partition
+                        .append(range)
+                        .ok_or(StorageError::NotAligned)?
+                }
+                storage_type::METADATA => {
+                    locations.metadata = locations
+                        .metadata
+                        .append(range)
+                        .ok_or(StorageError::NotAligned)?
+                }
+                _ => (),
+            };
+        }
+        if locations.partition.is_empty() || locations.metadata.is_empty() {
+            Err(StorageError::CustomError)
+        } else {
+            Ok(locations)
+        }
+    }
+
+    fn is_page_aligned(&self, x: usize) -> bool {
+        is_aligned(self.page_size, x)
     }
 }
 
-fn find_slice<'a>(
-    slices: &'a [&'a [u8]],
-    mut start: usize,
-    length: usize,
-) -> StorageResult<&'a [u8]> {
-    for slice in slices {
-        if start >= slice.len() {
-            start -= slice.len();
-            continue;
+impl UpgradeStorage for SyscallUpgradeStorage {
+    fn read_partition(&self, offset: usize, length: usize) -> StorageResult<&[u8]> {
+        if self
+            .partition
+            .contains_range(&ModRange::new(offset, length))
+        {
+            Ok(unsafe { core::slice::from_raw_parts(offset as *mut u8, length) })
```

Comment:
Ok so looking at this it really looks like the API for `read_partition` is that the offset is not an offset but an absolute position. This should be documented and possible renamed as `address`.

Or maybe this is a bug and the intent was:
```rust
if length <= partition.length && offset <= partition.length - length {
    Ok(from_raw_parts(partition.start + offset, length))
}
```

Reply:
Why is this `*mut u8` instead of `*const u8`? Is it some cast issue?

Reply:
Changed to `const`, just a copy&paste leftover.